lisa/microsoft/testsuites/cloud_hypervisor/ch_tests_tool.py

```diff
@@ -1338,6 +1426,707 @@ def _set_block_size_env_var(self, testcase: str) -> None:
+
+                current_mtu = int(mtu_match.group(1))
+
+                if current_mtu != expected_mtu:
+                    # Set MTU
+                    self.node.execute(
+                        f"sudo ip link set {primary_nic} mtu {expected_mtu}",
+                        shell=True,
+                        sudo=True,
+                    )
+
+    def _log_thermal_health(self, stage: str) -> None:
+        """
+        Log thermal health: sensors, throttle counts, frequency state.
+
+        Args:
+            stage: Execution stage (preflight, anchor_failure, suite_end)
+        """
+        self._log.info(f"=== Thermal Health [{stage}] ===")
+
+        # sensors output (if available)
+        result = self.node.execute(
+            "sensors | head -20",
+            shell=True,
+        )
+        if result.exit_code == 0 and result.stdout.strip():
+            for line in result.stdout.split("\n")[:10]:
+                if line.strip():
+                    self._log.info(f"  {line}")
+        else:
+            self._log.info("  Sensors: not available")
+
+        # Throttle counts
+        result = self.node.execute(
+            "for f in /sys/devices/system/cpu/cpu*/thermal_throttle/"
+            "*_throttle_count; do "
+            '[ -f "$f" ] && echo "$f: $(cat $f)"; done | head -5',
+            shell=True,
+        )
+        if result.exit_code == 0 and result.stdout.strip():
+            self._log.info(f"  Throttle counts: {result.stdout.strip()}")
+        else:
+            self._log.info("  Throttle counts: not available")
+
+        # CPU frequency (sample first 4 CPUs)
+        result = self.node.execute(
+            "for cpu in 0 1 2 3; do "
+            "f=/sys/devices/system/cpu/cpu$cpu/cpufreq/scaling_cur_freq; "
+            '[ -f "$f" ] && echo "CPU$cpu: $(cat $f) kHz"; done',
+            shell=True,
+        )
+        if result.exit_code == 0 and result.stdout.strip():
+            self._log.info(f"  Frequencies: {result.stdout.strip()}")
+
+    def _run_warmup_for_test(self, testcase: str) -> None:
+        """
+        Run 20-30s warmup before test measurement (perf-stable profile).
+        Extended to 30-45s for MQ random-write tests.
+
+        Uniform warmup pattern:
+        - Block tests: O_DIRECT dd reads (wake up NVMe queues)
+        - Network tests: nc or ping (wake up network stack)
+        - Other tests: CPU/mem warmup with dd
+
+        Cache policy:
+        - O_DIRECT tests: no cache drops (test uses direct I/O)
+        - Buffered writes: single drop before warmup only
+        - Reads: consistent hot/cold policy (no mixed)
+        """
+        # Base warmup duration
+        warmup_seconds = self.perf_warmup_seconds
+
+        # Extend for MQ random-write tests (30-45s helps SSD write path settle)
+        if self._is_mq_random_write_test(testcase):
+            warmup_seconds = max(warmup_seconds, self.MQ_WARMUP_MIN_SECONDS)
+
+        if warmup_seconds <= 0:
```

Comment:
Is there any case where this would be true?

Reply:
The if warmup_seconds <= 0 check is purely defensive, to handle:
Any future test harness override where warm-up is disabled globally (tool.perf_warmup_seconds = 0),
Or developer/debug runs that skip perf stabilization (for faster iteration).